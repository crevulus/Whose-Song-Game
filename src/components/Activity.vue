<template>
  <div class="activity flex flex-wrap mx-auto" style="max-width: 1280px">
    <div class="flex flex-wrap sm:flex-no-wrap w-full lg:w-2/3" style="max-width: 650px">
      <div class="w-panel mx-auto p-2">
        <h4 class="font-semibold">Whose song is this?</h4>
        <iframe
          v-if="currentSong.trackId"
          class="rounded-sm"
          :src="`https://open.spotify.com/embed/track/${currentSong.trackId}`"
          width="300"
          height="380"
          frameborder="0"
          allowtransparency="true"
          allow="encrypted-media"
        ></iframe>
      </div>
      <div class="w-panel mx-auto p-2">
        <PlayerSelectionList
          v-if="!isSongOwner"
          :users="this.users"
          :userId="this.deviceId"
          :guess="currentGuess"
          :currentSong="currentSong"
        />
        <UserGuesses v-if="isSongOwner" :users="usersGuessList" :userId="this.deviceId" />
      </div>
    </div>
    <div
      class="flex flex-col p-2 bg-purple-700 rounded-xl rounded-r-none lg:rounded-r-xl lg:px-8 lg:mr-3 py-16 absolute right-0 top-0 h-full"
      :class="{'lg:rounded-xl px-4': collapsed, 'px-8': !collapsed}"
    >
      <!-- Toggle collapse participants panel -->
      <div class="lg:hidden absolute left-0" style="top: 50%; transform: translateY(-50%);">
        <button v-on:click="collapsed = !collapsed">
          <svg
            v-if="collapsed"
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            stroke="#fff"
            stroke-width="2"
            stroke-linecap="round"
            stroke-linejoin="round"
            class="feather feather-chevron-left focus:outline-none"
          >
            <polyline points="15 18 9 12 15 6" />
          </svg>
          <svg
            v-if="!collapsed"
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            stroke="#fff"
            stroke-width="2"
            stroke-linecap="round"
            stroke-linejoin="round"
            class="feather feather-chevron-right"
          >
            <polyline points="9 18 15 12 9 6" />
          </svg>
        </button>
      </div>

      <div
        id="participants"
        class="lg:block flex-grow mb-4"
        style="overflow-y: auto; width: 290px"
        :class="{'hidden': collapsed, 'is-host': isHost}"
      >
        <p class="text-white font-semibold">Participants</p>
        <Participants
          :songOwner="currentSong.userId"
          :participants="participants"
          :isSongOwner="isSongOwner"
          :hasGuessed="hasGuessed"
        />
      </div>
      <div v-if="isHost" class="mt-auto lg:block" :class="{'hidden': collapsed}">
        <p class="text-white font-semibold">ðŸ•¹Host controls</p>
        <t-button
          class="w-full mb-4"
          v-if="songs.length > 0"
          variant="warning"
          @click="showNextSong"
        >Skip Song</t-button>
        <t-button
          class="w-full mb-4"
          variant="warning"
          v-else
          @click="endActivityInstanceMutation"
        >Skip to Results</t-button>
        <t-button class="w-full" variant="warning" @click="endActivityInstanceMutation">End Game</t-button>
      </div>
    </div>

    <!-- Component that shows instance data in tables -->
    <ActivityDebugger
      v-if="this.showDebugger"
      :users="this.users"
      :score="this.score"
      :currentSong="this.currentSong"
      :songs="this.songs"
      :playedSongs="this.playedSongs"
      :guesses="this.guesses"
    />
  </div>
</template>
<script>
import { API, graphqlOperation } from "aws-amplify";
import * as mutations from "@/graphql/mutations";
import * as subscriptions from "@/graphql/subscriptions";
import * as queries from "@/graphql/queries";
import commonMethods from "@/mixins/commonMethods";
import ActivityDebugger from "@/components/ActivityDebugger";
import PlayerSelectionList from "@/components/PlayerSelectionList";
import Participants from "@/components/Participants";
import UserGuesses from "@/components/UserGuesses";

export default {
  name: "Activity",
  components: {
    ActivityDebugger,
    PlayerSelectionList,
    Participants,
    UserGuesses
  },
  mixins: [commonMethods],
  data() {
    return {
      currentSong: {},
      songs: [],
      playedSongs: [],
      score: [],
      guesses: [],
      showDebugger: false,
      currentGuess: null,
      guessedList: [],
      collapsed: true
    };
  },
  computed: {
    participants() {
      return this.users.map(user => ({
        name: user.name,
        hasGuessed: this.userHasGuessed(user.userId),
        userId: user.userId
      }));
    },
    hasNextSong() {
      return this.songs.length > 0;
    },
    isSongOwner() {
      return this.deviceId === this.currentSong.userId;
    },
    hasGuessed() {
      return this.userHasGuessed(this.deviceId);
    },
    usersGuessList() {
      return this.users.map(user => {
        const guess = this.guesses.find(
          guess =>
            guess.userId === user.userId &&
            guess.trackOwnerId === this.currentSong.userId
        );
        return {
          name: user.name,
          userId: user.userId,
          hasGuessed: !!guess,
          isCorrect: guess && guess.selectedUserId === guess.trackOwnerId
        };
      });
    }
  },
  watch: {
    guessedList: function() {
      if (!this.isHost) return;

      const shouldUpdate = this.guessedList.length === this.users.length - 1;
      if (shouldUpdate && this.songs.length) {
        this.showNextSong();
        return;
      }
      if (shouldUpdate) {
        API.graphql(
          graphqlOperation(mutations.endActivityInstance, {
            activityInstanceId: this.activityInstanceId
          })
        );
      }
    }
  },
  created() {
    this.getActivityInstanceQuery();
    this.getActivityInstanceData();
    this.updatedActivityInstanceDataSubscription();
  },
  methods: {
    endActivityInstanceMutation() {
      API.graphql(
        graphqlOperation(mutations.endActivityInstance, {
          activityInstanceId: this.activityInstanceId
        })
      );
    },
    getActivityInstanceData() {
      API.graphql(
        graphqlOperation(queries.whoseSongGetActivityInstanceData, {
          activityInstanceId: this.activityInstanceId,
          userId: this.deviceId
        })
      ).then(response => {
        const data = response.data.whoseSongGetActivityInstanceData;
        this.setVariables(data);
      });
    },
    showNextSong() {
      API.graphql(
        graphqlOperation(mutations.whoseSongUpdateActivityInstanceData, {
          action: "setNextSong",
          activityInstanceId: this.activityInstanceId,
          userId: this.deviceId
        })
      );
    },
    updatedActivityInstanceDataSubscription() {
      API.graphql(
        graphqlOperation(subscriptions.whoseSongUpdatedActivityInstanceData, {
          activityInstanceId: this.activityInstanceId
        })
      ).subscribe(response => {
        const data = response.value.data.whoseSongUpdatedActivityInstanceData;
        this.setVariables(data);
      });
    },
    setVariables(data) {
      this.currentSong = data.currentSong;
      this.playedSongs = data.playedSongs;
      this.score = data.score;
      this.songs = data.songs;
      this.guesses = data.guesses;
      this.guessedList = this.guesses.filter(guess => {
        const { trackId, userId } = this.currentSong;
        return guess.trackId === trackId && guess.trackOwnerId === userId;
      });
      this.currentGuess = this.guessedList.find(
        guess => guess.userId === this.deviceId
      );
    },
    userHasGuessed(userId) {
      return !!this.guesses.find(
        guess =>
          guess.userId === userId &&
          guess.trackOwnerId === this.currentSong.userId
      );
    }
  }
};
</script>
<style>
participants:not(.is-host) {
  height: calc(100% - 8rem);
}
</style>
