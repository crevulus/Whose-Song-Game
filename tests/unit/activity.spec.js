import { shallowMount, config, createLocalVue } from "@vue/test-utils";
import Activity from "@/components/Activity";
import VueTailwind from "vue-tailwind";

const mockMixin = {
  commonMethods: {}
}

const localVue = createLocalVue();
localVue.use(VueTailwind);
localVue.mixin(mockMixin);

config.mocks["$gtm"] = { trackView: function (name, path) { } };
config.mocks["$route"] = { params: { activityInstanceId: '123' } };
// config.mocks['$router'] = { resolve: arg => { href: 'url' } };


describe('Activity Component', () => {


  const methods = {
    getActivityQuery() {
      return;
    },
    getActivityInstanceData() {
      return;
    },
    updatedActivityInstanceSubscription() {
      return;
    },
    updatedActivityInstanceDataSubscription() {
      return;
    }
  }

  const users = [
    {
      userId: '123',
      name: 'Foo'
    },
    {
      userId: '456',
      name: 'Bar'
    }
  ]

  const data = {
    activityInstanceId: '123456789',
    isHost: false,
    users,
    hostId: '456',
    status: 'started',
    deviceId: '789',
    showDebugger: false
  }

  const wrapper = shallowMount(Activity, {
    localVue,
    methods,
    data() {
      return data;
    },
    mixins: mockMixin
  })

  it('testing setVariables method', () => {
    const tracks = [{
      id: 123,
      name: 'foo',
      artists: [{ name: 'Tell' }],
      album: { name: 'Mitch', images: { 2: { url: '123' } } }
    },
    {
      id: 1234,
      name: 'foo',
      artists: [{ name: 'Tell' }],
      album: { name: 'Mitch', images: { 2: { url: '123' } } }
    }];

    const guesses = [
      {
        trackId: '789',
        userId: '678',
        selectedUserId: '678',
        trackOwnerId: '123'
      }
    ]

    const data2 = {
      currentSong: tracks[0],
      playedSong: tracks,
      score: [1, 6, 3],
      songs: tracks,
      guesses: guesses
    }

    wrapper.vm.setVariables(data2);

    expect(wrapper.vm.currentSong).toBe(tracks[0]);
  })

  test('renders correctly (snapshot)', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
