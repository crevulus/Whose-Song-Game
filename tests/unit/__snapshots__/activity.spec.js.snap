// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity Component renders correctly (snapshot) 1`] = `
<div
  class="activity flex flex-wrap mx-auto"
  style="max-width: 1280px;"
>
  <div
    class="flex flex-wrap sm:flex-no-wrap w-full lg:w-2/3"
  >
    <div
      class="w-full mx-auto sm:w-1/2 p-2"
    >
      <h4
        class="font-semibold"
      >
        Whose song is this?
      </h4>
       
      <iframe
        allow="encrypted-media"
        allowtransparency="true"
        class="rounded-sm"
        frameborder="0"
        height="380"
        src="https://open.spotify.com/embed/track/undefined"
        width="300"
      />
    </div>
     
    <div
      class="w-full mx-auto sm:flex-grow p-2"
    >
      <h4
        class="font-semibold"
      >
        Select a name and confirm
      </h4>
       
      <playerselectionlist-stub
        currentsong="[object Object]"
        userid="789"
        users="[object Object],[object Object]"
      />
    </div>
  </div>
   
  <div
    class="w-full mx-auto lg:w-1/3 p-2"
  >
    <h1>
      Host controls
    </h1>
     
    <!---->
     
    <t-button-stub
      activeclass="router-link-active"
      baseclass="border block rounded inline-flex items-center justify-center"
      dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600"
      defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500"
      defaultsizeclass="px-6 py-3"
      disabledclass="cursor-not-allowed opacity-75"
      exactactiveclass="router-link-exact-active"
      largesizeclass="px-8 py-4 text-lg"
      primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600"
      secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white"
      smallsizeclass="px-4 py-2 text-sm"
      successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600"
      tagname="button"
      tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600"
      type="button"
      variant="primary"
      warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600"
    >
      End Game
    </t-button-stub>
  </div>
   
  <!---->
</div>
`;
